### ISSUE TITLE
Firewall rule 'Allow DNS queries' does not take into account custom (VPN) DNS server

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#1183]
https://groups.google.com/d/msg/qubes-users/EPe4U0kF67A/CJfHD245KwAJ

> There appears to be a bug in Qubes firewall dns forwarding after
> 'qubes-setup-dnat-to-ns' is called to update the nat PR-QBS chain /and/
> there are 'Deny except...' whitelist rules added from a connected
> (downstream) appvm.
> 
> The destination addresses in the Qubes-generated dpt:domain rules in the
> FORWARD chain remain their old direct-to-netvm values instead of getting
> updated to the vpn addresses. So those nat'd dns packets keep hitting the
> 'reject-with icmp-host-prohibited' rule since they are never matched and
> accepted above.
> 
> The firewall might need to be fixed so that whatever new dns addresses are
> picked up by 'qubes-setup-dnat-to-ns', the dpt:domain rules in FORWARD are
> also changed to look for those addresses.

Reported by theman, cprise.

**Comment by 3hhh** [comment]
Still valid in 3.2.

**Comment by 3hhh** [comment]
workaround:
iptables-save | sed -r 's/10\.137\.2\.1\/32/[VPN-SERVER-DNS-IP]\/32/g'|iptables-restore fixes the issue until the qubes firewall refreshes (can be persisted using /rw/config/qubes-firewall-user-script)

**Comment by marmarek** [comment]
This is already handled in Qubes 4.0 (as part of #1815 ). But it isn't feasible for backport to R3.2 unfortunately.

**Comment by tasket** [comment]
Comprehensive regex to fix the FORWARD chain, per @3hhh suggestion:
```
sed -ri 's/(-A FORWARD.+ -d )10\.137\.[0-9]+\.1(\/32 .+--dport 53 -j ACCEPT)$/\1'$DNS1'\2/'
sed -ri 's/(-A FORWARD.+ -d )10\.137\.[0-9]+\.254(\/32 .+--dport 53 -j ACCEPT)$/\1'$DNS2'\2/'

```

This could be added to both qubes-firewall-user-script and qubes-vpn-handler.

**Comment by tasket** [comment]
Should also investigate possibility of re-populating /var/run/qubes/qubes-ns with new DNS addresses.

**Comment by tasket** [comment]
Committed a fix for this in my VPN project. This can be added to the VPN doc as well.
(It also removes one editing step.)

**Comment by tasket** [comment]
I'm thinking this could be fixed in R3.2.x and maybe the best overall way is to use an additional file like /var/run/qubes/qubes-tun-ns to hold the tunnel's DNS addresses. This can be used by the firewall if it exists.

Another way is to put the `sed` code from my script into qubes-setup-dnat-to-ns, which is even triggered by Network Manager for VPN connections. Doing it this way would make stuff 'just work' with either scripted or NM connection methods.

**Comment by 3hhh** [comment]
The sed solution was a mere hack and should never make it into any proper OS design.

Moreover it only works if you switch the "Allow DNS queries" checkbox even if you allow all network traffic (which is different from the standard Qubes OS behavior).

**Comment by tasket** [comment]
Correct, it only works under that condition but that's because the purpose is to prevent unnecessary blocking of DNS. Using `sed` seems hackish, but I noticed this elsewhere in Qubes so figured it wasn't too bad.

OTOH, the other solution just imitates what is done with /var/run/qubes/qubes-ns, seems simpler.

**Comment by 3hhh** [comment]
This is somewhat fixed in 4.0rc1 as the dynamic iptables rules in proxyVMs were removed entirely, i.e. custom DNS servers should work.

Only drawback: **Local** DNS servers are affected by #3051

**Comment by marmarek** [comment]
>  the dynamic iptables rules in proxyVMs were removed entirely, i.e. custom DNS servers should work.

This is not true. There are moved into nftables. And they do support custom dns (based on what you have in /etc/resolv.conf). See `specialtarget=dns` [here](https://www.qubes-os.org/doc/vm-interface/),
